portfolio/src/main/java/com/google/sps/data/ArrayListTimeRangeGroup.java

```diff
@@ -0,0 +1,189 @@
+      // after the merging.
+      if (lastExaminedTimeRange.overlaps(currentRange)) {
+        lastExaminedTimeRange = mergeTwoTimeRanges(currentRange, lastExaminedTimeRange);
+      } else if (currentRange.end().isAfter(lastExaminedTimeRange.end())) {
+        // This is the case that lastExaminedTimeRange and the current time range do not overlap.
+        // If the current range from the original list ends after the last examined time range,
+        // add the time range pointed to by lastExaminedTimeRange to the new list of all ranges, and
+        // change the lastExaminedTimeRange to point to the current range.
+        newTimeRanges.add(lastExaminedTimeRange);
+        lastExaminedTimeRange = currentRange;
+      } else {
+        // If lastExaminedTimeRange is later than the current range,
+        // add the current range to the new list of all ranges.
+        // The lastExaminedTimeRange pointer shouldn't move.
+        newTimeRanges.add(currentRange);
+      }
+
+      // If current time range is the last element in the allTimeRanges list,
+      // add the last examined time range to the new list.
+      if (i == allTimeRanges.size() - 1) {
+        newTimeRanges.add(lastExaminedTimeRange);
+      }
+    }
+
+    // Finally, set the global variable allTimeRanges to this newly built list of time ranges.
+    allTimeRanges = newTimeRanges;
+  }
+
+  /**
+   * Helper method for merging two time ranges. This method is package-private so that it can be
+   * tested.
+   */
+  static TimeRange mergeTwoTimeRanges(TimeRange a, TimeRange b) {
+    if (!a.overlaps(b)) {
+      throw new IllegalArgumentException("Merging two time ranges that do not overlap is invalid");
+    }
+
+    Instant newTimeRangeStart;
+    Instant newTimeRangeEnd;
+
+    // The new time range after merging should have the earlier start time
+    // and the later end time among the two overlapping time ranges.
+    if (a.start().isBefore(b.start())) {
+      newTimeRangeStart = a.start();
+    } else {
+      newTimeRangeStart = b.start();
+    }
+
+    if (a.end().isBefore(b.end())) {
+      newTimeRangeEnd = b.end();
+    } else {
+      newTimeRangeEnd = a.end();
+    }
+
+    return TimeRange.fromStartEnd(newTimeRangeStart, newTimeRangeEnd);
+  }
+
+  /** Returns an iterator for the list of all time ranges. */
+  public Iterator<TimeRange> iterator() {
+    return allTimeRanges.iterator();
+  }
+
+  /**
+   * Checks if a time range exists in the collection. For example, if [3:00 - 4:00] is in the
+   * collection, [3:00 - 3:30] is considered to exist as a time range in the collection. This method
+   * uses binary search to find the time ranges whose start time is before the target range's start
+   * and whose end time is after the target range's end. Then the method calls contains to see if
+   * the target range is contained within this current range.
+   */
+  public boolean hasTimeRange(TimeRange timeRangeToCheck) {
```

Comment:
I think this will crash with an `IndexOutOfBoundsException` if the group is empty since you'll end up with `start == end == 0`, skip the loop then call `allTimeRanges.get(0)`?

Can you special case the empty set to return `false` and then add a test?

Reply:
I see. Updated.